## Question 1: What is Logistic Regression, and how does it differ from Linear Regression?


###Logistic Regression
- **Logistic Regression** is a statistical method used for **classification problems**.  
- It predicts the **probability** of a data point belonging to a particular class (e.g., Yes/No, Spam/Not Spam, Fraud/Not Fraud).  
- Instead of fitting a straight line, it uses a **sigmoid (logistic) function** to map predicted values between **0 and 1**.

### Logistic Function (Sigmoid Function):
P(Y=1|X) = 1/(1 + e^-(beta_0 + beta_1 X))


Where:
- P(Y=1|X)= Probability of the outcome being 1  
- (beta_0,beta_1) = Model coefficients  
- (e) = Base of the natural logarithm  


### Linear Regression
- **Linear Regression** is used for predicting a **continuous outcome** (e.g., house price, salary, sales).  
- It assumes a **linear relationship** between independent variable(s) and the dependent variable.  

### Linear Regression Equation:

Y = beta_0 + beta_1 X + epsilon


Where:
-  Y  = Predicted continuous value  
- beta_0, beta_1  = Model coefficients  
- epsilon  = Error term  

### Key Differences Between Logistic & Linear Regression

| Feature                  | Linear Regression                            | Logistic Regression                           |
|---------------------------|----------------------------------------------|----------------------------------------------|
| **Output**               | Continuous value (e.g., 2.5, 100.7, etc.)   | Probability between 0 and 1                  |
| **Use Case**             | Regression problems (predicting numbers)     | Classification problems (Yes/No, 0/1, etc.) |
| **Function**             | Straight line                                | S-shaped sigmoid curve                       |
| **Interpretation**       | Direct predicted values                      | Probability of belonging to a class          |
| **Assumption**           | Linear relationship                         | Log-odds relationship with predictors        |

###Example
- **Linear Regression**: Predicting house price based on area.  
- **Logistic Regression**: Predicting whether a customer will buy a product (Yes=1, No=0) based on income.  



---
## Question 2: Explain the role of the Sigmoid function in Logistic Regression

###What is the Sigmoid Function?
The **Sigmoid function** is a mathematical function that maps any real-valued number into a value between **0 and 1**.  
It is defined as:

sigma(z) = 1/{1 + e^{-z}}

Where:
z = beta_0 + beta_1 X_1 + beta_2 X_2 + ... + beta_n X_n (the linear combination of features and weights)

### Role in Logistic Regression
1. **Converts Linear Output to Probability**  
   - Logistic Regression first computes a **linear equation** like Linear Regression.  
   - The Sigmoid function then transforms this linear output into a **probability value (0–1)**.  

2. **Classification Decision**  
   - If the probability  P(Y=1|X) > 0.5 , predict **Class 1**.  
   - Otherwise, predict **Class 0**.  

3. **Handles Non-linearity**  
   - The S-shape of the Sigmoid allows Logistic Regression to model probabilities effectively, even when the linear regression output ranges from (-infty) to (+infty).  

### Example
Suppose the model predicts a linear output ( z = 2.3 ).  
Using the Sigmoid function:


P(Y=1|X) = 1/{1 + e^{-2.3}} = approx 0.91


This means there is a **91% probability** that the instance belongs to **Class 1**.




---
# Question 3: What is Regularization in Logistic Regression and why is it needed?

### Regularization:
**Regularization** is a technique used in Logistic Regression (and other ML models) to **prevent overfitting** by adding a **penalty term** to the loss function.  

- Logistic Regression normally minimizes the **Log Loss (Cross-Entropy Loss)**.  
- With Regularization, an additional penalty term is added based on the **magnitude of coefficients (weights)**.  

This discourages the model from assigning **too large weights** to features.

### Types of Regularization
1. **L1 Regularization (Lasso)**  
   - Adds the **absolute values** of coefficients to the loss function.  
   - Encourages **sparsity** : some coefficients become **exactly 0** → useful for **feature selection**.  
   - Penalty term:  
   lambda sum_{j=1} to n | beta_j |
   

2. **L2 Regularization (Ridge)**  
   - Adds the **squared values** of coefficients to the loss function.  
   - Shrinks coefficients but rarely makes them exactly 0.  
   - Useful when features are highly correlated.  
   - Penalty term:  
   lambda sum_{j=1}to n (beta_j)^2

3. **Elastic Net Regularization**  
   - Combines **L1 and L2 penalties**.  
   - Balances between **feature selection (L1)** and **coefficient shrinkage (L2)**.  

### Need
- **Prevents Overfitting**: Ensures the model does not memorize training data.  
- **Improves Generalization**: Helps model perform better on unseen data.  
- **Handles Multicollinearity**: L2 regularization reduces the impact of correlated features.  
- **Feature Selection**: L1 regularization automatically removes irrelevant features.  



---
## Question 4: What are some common evaluation metrics for classification models, and why are they important?

###Evaluation Metrics  Importantance
- Classification models need to be evaluated to check **how well they perform**.  
- Different metrics highlight **different aspects of performance** (accuracy, precision, recall, etc.).  
- Relying on a single metric (like accuracy) can be **misleading**, especially with **imbalanced datasets**.  


### Common Evaluation Metrics

### 1. **Accuracy**
- Proportion of correctly classified samples.  

Accuracy = {TP + TN} / {TP + TN + FP + FN}

- Useful when classes are **balanced**.  
- Misleading when dataset is **imbalanced**.  

### 2. **Precision**
- Out of all predicted positives, how many are actually positive.  

Precision = {TP} / {TP + FP}
- Important in applications where **false positives are costly** (e.g., spam detection).  


### 3. **Recall (Sensitivity or True Positive Rate)**
- Out of all actual positives, how many are correctly predicted.  

Recall = {TP} / {TP + FN}

- Important in cases where **false negatives are costly** (e.g., cancer detection).  

### 4. **F1 Score**
- Harmonic mean of Precision and Recall.  

F1 = 2 * {{Precision * Recall}/{Precision + Recall}}

- Useful when there is a need to balance **precision and recall**.  



### 5. **ROC Curve and AUC (Area Under Curve)**
- **ROC Curve**: Plots **True Positive Rate (Recall)** vs. **False Positive Rate**.  
- **AUC**: Measures overall ability of the model to distinguish between classes.  
- AUC close to 1 → better performance.  

### 6. **Confusion Matrix**
- A table showing counts of **True Positives (TP), True Negatives (TN), False Positives (FP), False Negatives (FN)**.  
- Provides a **complete picture** of classification results.  

---  


In [131]:
# Question 5: Write a Python program that loads a CSV file into a Pandas DataFrame,
# splits into train/test sets, trains a Logistic Regression model, and prints its accuracy.
# (Use Dataset from sklearn package)

In [132]:
import pandas as pd
import numpy as np
import seaborn as sns
import warnings
warnings.filterwarnings(action='ignore')
from sklearn.datasets import load_breast_cancer # Loading Diabetes Data from sklearn datasets
breast_cancer_data=load_breast_cancer()

In [133]:
df=pd.DataFrame(breast_cancer_data.data,columns=breast_cancer_data.feature_names)
df["cancer"]=breast_cancer_data.target
df.sample(8)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
507        11.06         17.12           71.25      366.5          0.11940   
15         14.54         27.54           96.73      658.8          0.11390   
276        11.33         14.16           71.79      396.6          0.09379   
349        11.95         14.96           77.23      426.7          0.11580   
0          17.99         10.38          122.80     1001.0          0.11840   
18         19.81         22.15          130.00     1260.0          0.09831   
49         13.49         22.30           86.91      561.0          0.08752   
251        11.50         18.45           73.28      407.4          0.09345   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
507           0.10710        0.040630             0.042680         0.1954   
15            0.15950        0.163900             0.073640         0.2303   
276           0.03872        0.001487             0.003333         0.1954   
349           0.12060        0.011710             0.017870         0.2459   
0             0.27760        0.300100             0.147100         0.2419   
18            0.10270        0.147900             0.094980         0.1582   
49            0.07698        0.047510             0.033840         0.1809   
251           0.05991        0.026380             0.020690         0.1834   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
507                 0.07976  ...          20.74            76.08       411.1   
15                  0.07077  ...          37.13           124.10       943.2   
276                 0.05821  ...          18.99            77.37       458.0   
349                 0.06581  ...          17.72            83.09       496.2   
0                   0.07871  ...          17.33           184.60      2019.0   
18                  0.05395  ...          30.88           186.80      2398.0   
49                  0.05718  ...          31.82            99.00       698.8   
251                 0.05934  ...          22.46            83.12       508.9   

     worst smoothness  worst compactness  worst concavity  \
507            0.1662            0.20310         0.125600   
15             0.1678            0.65770         0.702600   
276            0.1259            0.07348         0.004955   
349            0.1293            0.18850         0.031220   
0              0.1622            0.66560         0.711900   
18             0.1512            0.31500         0.537200   
49             0.1162            0.17110         0.228200   
251            0.1183            0.10490         0.081050   

     worst concave points  worst symmetry  worst fractal dimension  cancer  
507               0.09514          0.2780                  0.11680       1  
15                0.17120          0.4218                  0.13410       0  
276               0.01111          0.2758                  0.06386       1  
349               0.04766          0.3124                  0.07590       1  
0                 0.26540          0.4601                  0.11890       0  
18                0.23880          0.2768                  0.07615       0  
49                0.12820          0.2871                  0.06917       1  
251               0.06544          0.2740                  0.06487       1  

[8 rows x 31 columns]

In [134]:
x=df.drop(columns=['cancer'])
y=df["cancer"]

In [135]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)


In [136]:
from sklearn.linear_model import LogisticRegression
model=LogisticRegression(max_iter=10000)
model.fit(x_train,y_train)

LogisticRegression(max_iter=10000)

In [137]:
y_pred=model.predict(x_test)

In [138]:
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred))

0.9473684210526315


In [139]:
# Question 6: Write a Python program to train a Logistic Regression model using L2
# regularization (Ridge) and print the model coefficients and accuracy.

In [140]:
data=sns.load_dataset('iris')
data.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [141]:
data["species"].unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [142]:
data["species"]=data["species"].map({'setosa':0, 'versicolor':1, 'virginica':2})

In [143]:
x=data.drop('species',axis=1)
y=data['species']

In [144]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)

In [145]:
l2_model=LogisticRegression(penalty='l2')
l2_model

LogisticRegression()

In [146]:
l2_model.fit(x_train,y_train)

LogisticRegression()

In [147]:
l2_model.feature_names_in_

array(['sepal_length', 'sepal_width', 'petal_length', 'petal_width'],
      dtype=object)

In [148]:
coef=pd.DataFrame(l2_model.coef_.T,columns=['class 0 coef','class 2 coef','class 2 coef'])
coef["features"]=l2_model.feature_names_in_
coef  # table for coefficients for each class for each feature

class 0 coef  class 2 coef  class 2 coef      features
0     -0.431713      0.618185     -0.186472  sepal_length
1      0.823447     -0.428154     -0.395293   sepal_width
2     -2.351192     -0.205960      2.557152  petal_length
3     -0.969380     -0.829523      1.798903   petal_width

In [149]:
print(accuracy_score(y_test,l2_model.predict(x_test)))

0.9666666666666667


In [150]:
# Question 7: Write a Python program to train a Logistic Regression model for multiclass
# classification using multi_class='ovr' and print the classification report.
# (Use Dataset from sklearn package)

In [151]:
from sklearn.datasets import make_classification
x,y=make_classification(n_classes=3,n_features=10,n_informative=5,n_redundant=5,n_samples=1000)

In [152]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)

In [153]:
multiclass_model=LogisticRegression(multi_class='ovr')
multiclass_model

LogisticRegression(multi_class='ovr')

In [154]:
multiclass_model.fit(x_train,y_train)

LogisticRegression(multi_class='ovr')

In [155]:
from sklearn.metrics import classification_report
print(classification_report(y_test,multiclass_model.predict(x_test)))

              precision    recall  f1-score   support

           0       0.89      0.81      0.85        81
           1       0.69      0.80      0.74        56
           2       0.75      0.73      0.74        63

    accuracy                           0.79       200
   macro avg       0.78      0.78      0.78       200
weighted avg       0.79      0.79      0.79       200



In [156]:
# Question 8: Write a Python program to apply GridSearchCV to tune C and penalty
# hyperparameters for Logistic Regression and print the best parameters and validation
# accuracy.


In [157]:
from sklearn.datasets import make_classification
x,y=make_classification(n_classes=2,n_features=10,n_informative=5,n_redundant=5,n_samples=1000)

In [158]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=1)

In [159]:
from sklearn.model_selection import GridSearchCV
params={'C':[0.5,1.0,2.0,1.5,1.2,3],'penalty':['l1','l2','elasticnet']}
gridsearch=GridSearchCV(estimator=LogisticRegression(),param_grid=params,cv=5)
gridsearch

GridSearchCV(cv=5, estimator=LogisticRegression(),
             param_grid={'C': [0.5, 1.0, 2.0, 1.5, 1.2, 3],
                         'penalty': ['l1', 'l2', 'elasticnet']})

In [160]:
gridsearch.fit(x_train,y_train)

GridSearchCV(cv=5, estimator=LogisticRegression(),
             param_grid={'C': [0.5, 1.0, 2.0, 1.5, 1.2, 3],
                         'penalty': ['l1', 'l2', 'elasticnet']})

In [163]:
gridsearch.best_params_

{'C': 0.5, 'penalty': 'l2'}

In [164]:
print(accuracy_score(y_test,gridsearch.predict(x_test)))

0.845


In [ ]:
# Question 9: Write a Python program to standardize the features before training Logistic
# Regression and compare the model's accuracy with and without scaling.

In [7]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
import pandas as pd

breast_cancer_data = load_breast_cancer()

df = pd.DataFrame(breast_cancer_data.data, columns=breast_cancer_data.feature_names)
df['cancer'] = breast_cancer_data.target

x = df.drop(columns=['cancer'])
y = df['cancer']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=1)

In [3]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# train the model
model_unscaled = LogisticRegression(max_iter=1000,solver='newton-cg')
model_unscaled.fit(x_train, y_train)

y_pred_unscaled = model_unscaled.predict(x_test)

#accuracy
accuracy_unscaled = accuracy_score(y_test, y_pred_unscaled)
print(f"Accuracy without scaling: {accuracy_unscaled}")

Accuracy without scaling: 0.9473684210526315





Standardize the features of the training and testing sets using StandardScaler and display the head of the scaled training data.



In [8]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

x_train_scaled = pd.DataFrame(x_train_scaled, columns=x_train.columns)
x_test_scaled = pd.DataFrame(x_test_scaled, columns=x_test.columns)

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

model_scaled = LogisticRegression(max_iter=1000)
model_scaled.fit(x_train_scaled, y_train)

y_pred_scaled = model_scaled.predict(x_test_scaled)

# accuracy
accuracy_scaled = accuracy_score(y_test, y_pred_scaled)
print(f"Accuracy with scaling: {accuracy_scaled}")

Accuracy with scaling: 0.9736842105263158


## Compare accuracies




In [6]:
print(f"Accuracy without scaling: {accuracy_unscaled}")
print(f"Accuracy with scaling: {accuracy_scaled}")

if accuracy_scaled > accuracy_unscaled:
    print("Feature scaling improved the accuracy of the Logistic Regression model.")
elif accuracy_scaled < accuracy_unscaled:
    print("Feature scaling slightly decreased the accuracy of the Logistic Regression model.")
else:
    print("Feature scaling did not significantly change the accuracy of the Logistic Regression model.")

Accuracy without scaling: 0.9473684210526315
Accuracy with scaling: 0.9736842105263158
Feature scaling improved the accuracy of the Logistic Regression model.



### Findings

*   The accuracy of the Logistic Regression model trained on unscaled data was approximately 0.947. A `ConvergenceWarning` was observed during training without scaling.
*   After standardizing the features, the accuracy of the Logistic Regression model increased to approximately 0.974.


*   Feature scaling, specifically standardization, improved the accuracy of the Logistic Regression model for this dataset and resolved the convergence issue observed with unscaled data.
*   For future modeling tasks on this dataset, it is recommended to include feature scaling as a standard preprocessing step.


In [ ]:
# Question 10: Imagine you are working at an e-commerce company that wants to
# predict which customers will respond to a marketing campaign. Given an imbalanced
# dataset (only 5% of customers respond), describe the approach you’d take to build a
# Logistic Regression model — including data handling, feature scaling, balancing
# classes, hyperparameter tuning, and evaluating the model for this real-world business
# use case. and write code as well.


## Simulate imbalanced data


Creating a synthetic imbalanced dataset that mimics the e-commerce scenario where only a small percentage of customers respond to the marketing campaign.


In [9]:
from sklearn.datasets import make_classification

x_imb, y_imb = make_classification(n_samples=10000, n_features=20, n_informative=15, n_redundant=5, weights=[0.95], n_classes=2, random_state=42)

print("Shape of x_imb:", x_imb.shape)
print("Shape of y_imb:", y_imb.shape)

Shape of x_imb: (10000, 20)
Shape of y_imb: (10000,)


## Split data



In [10]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_imb, y_imb, test_size=0.2, random_state=42)

print("Shape of x_train:", x_train.shape)
print("Shape of x_test:", x_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of x_train: (8000, 20)
Shape of x_test: (2000, 20)
Shape of y_train: (8000,)
Shape of y_test: (2000,)


## Feature scaling





Standardizing the features of the training and testing sets using StandardScaler .



In [11]:
from sklearn.preprocessing import StandardScaler
import pandas as pd

scaler = StandardScaler()

x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

x_train_scaled = pd.DataFrame(x_train_scaled, columns=[f'feature_{i}' for i in range(x_train_scaled.shape[1])])
x_test_scaled = pd.DataFrame(x_test_scaled, columns=[f'feature_{i}' for i in range(x_test_scaled.shape[1])])

print("Shape of x_train_scaled:", x_train_scaled.shape)
print("Shape of x_test_scaled:", x_test_scaled.shape)
display(x_train_scaled.head())

Shape of x_train_scaled: (8000, 20)
Shape of x_test_scaled: (2000, 20)


feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
0  -0.810497   0.460802  -0.706682  -0.020117  -0.127205   0.012224   
1  -1.184044   0.268064  -0.260163  -0.434954   1.991921  -0.426588   
2   0.329606   0.275134  -1.570498   1.123946   0.076762   0.659360   
3  -0.345200  -0.969794   0.186851  -1.452245  -1.925017  -0.300075   
4  -0.935159   0.471973  -3.268942   2.854145  -0.632131   0.031023   

   feature_6  feature_7  feature_8  feature_9  feature_10  feature_11  \
0   1.283515   0.220624   1.611099   1.919654   -0.172738   -1.499727   
1   1.411022   0.556895  -0.442309  -0.765006    0.073715    3.057018   
2   1.743980  -0.267959   1.274174  -1.142141   -1.199570   -0.511479   
3   1.924212  -0.772606   1.338695   0.821506    0.255355    0.886332   
4   3.012974  -0.133119   1.906585  -2.143793   -2.563478    0.505675   

   feature_12  feature_13  feature_14  feature_15  feature_16  feature_17  \
0    0.199597    0.558870   -0.400767   -2.587304    0.463738   -0.225950   
1    0.909006    0.562867   -1.066025    0.708151   -0.477664   -0.976225   
2   -0.113513    0.556014   -0.565186   -0.862821   -0.844964   -1.276942   
3   -0.091921    0.701328   -1.354146    1.217749   -0.398291    0.994612   
4   -1.581906   -1.173471    0.572834   -2.122287   -0.854604    0.459901   

   feature_18  feature_19  
0    1.872144   -0.018406  
1   -0.274853   -0.462239  
2    0.064875    0.794220  
3   -1.886465    0.656963  
4    3.006319    1.443810

## Handle class imbalance

Applying SMOTE (Synthetic Minority Over-sampling Technique) to balance the classes in the training set.


In [12]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state=42)

x_train_resampled, y_train_resampled = smote.fit_resample(x_train_scaled, y_train)

print("Shape of original training data (features):", x_train_scaled.shape)
print("Shape of original training data (target):", y_train.shape)
print("Shape of resampled training data (features):", x_train_resampled.shape)
print("Shape of resampled training data (target):", y_train_resampled.shape)


Shape of original training data (features): (8000, 20)
Shape of original training data (target): (8000,)
Shape of resampled training data (features): (15186, 20)
Shape of resampled training data (target): (15186,)


## Build and train model

Training a Logistic Regression model on the balanced training data.


In [13]:
from sklearn.linear_model import LogisticRegression

model_balanced = LogisticRegression(max_iter=1000)

model_balanced.fit(x_train_resampled, y_train_resampled)


LogisticRegression(max_iter=1000)

In [24]:
model_balanced

LogisticRegression(max_iter=1000)

## Hyperparameter tuning

Using GridSearchCV to tune the hyperparameters of the Logistic Regression model to find the best parameters for this imbalanced dataset.


In [14]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

param_grid = {'C': [0.1, 0.5, 1.0, 10.0, 100.0], 'penalty': ['l1', 'l2']}

grid_search = GridSearchCV(estimator=LogisticRegression(solver='liblinear', max_iter=1000), param_grid=param_grid, cv=5, scoring='roc_auc')

grid_search.fit(x_train_resampled, y_train_resampled)

print("Best parameters found: ", grid_search.best_params_)
print("Best cross-validation AUC: ", grid_search.best_score_)

Best parameters found:  {'C': 0.1, 'penalty': 'l1'}
Best cross-validation AUC:  0.9017855561091845


## Evaluate model

### Subtask:
Evaluate the performance of the tuned model on the *original*, *unscaled*, and *unbalanced* test set using appropriate metrics for imbalanced datasets (e.g., precision, recall, F1-score, ROC AUC).


**Reasoning**:
Evaluate the performance of the tuned model on the original, unscaled, and unbalanced test set using appropriate metrics for imbalanced datasets.



In [21]:
from sklearn.metrics import classification_report, roc_auc_score,confusion_matrix

y_pred_tuned = grid_search.predict(x_test_scaled)

print("Classification Report on Original Test Set:")
print(classification_report(y_test, y_pred_tuned))

y_pred_prob_tuned = grid_search.predict_proba(x_test_scaled)[:, 1]
roc_auc = roc_auc_score(y_test, y_pred_prob_tuned)
print(f"ROC AUC Score on Original Test Set: {roc_auc}")

Classification Report on Original Test Set:
              precision    recall  f1-score   support

           0       0.98      0.84      0.91      1868
           1       0.26      0.76      0.38       132

    accuracy                           0.84      2000
   macro avg       0.62      0.80      0.65      2000
weighted avg       0.93      0.84      0.87      2000

ROC AUC Score on Original Test Set: 0.8471992408020246


In [22]:
conf_matrix = confusion_matrix(y_test, y_pred_tuned)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[1578  290]
 [  32  100]]


## Analyze results

### Subtask:
Interpret the evaluation metrics in the context of the business problem.


**Reasoning**:
Analyze the printed classification report and ROC AUC score to interpret the model's performance on the imbalanced dataset in the context of the business problem.



In [16]:
print("Interpretation of Evaluation Metrics:")
print("\nClassification Report:")
print(classification_report(y_test, y_pred_tuned))

print(f"\nROC AUC Score: {roc_auc}")

print("\nAnalysis in the context of the e-commerce marketing campaign:")
print(f"- Precision for the minority class (responders, class 1): {classification_report(y_test, y_pred_tuned, output_dict=True)['1']['precision']:.2f}")
print(f"  - This metric indicates that when the model predicts a customer will respond, it is correct {classification_report(y_test, y_pred_tuned, output_dict=True)['1']['precision']:.2f}% of the time.")
print(f"- Recall for the minority class (responders, class 1): {classification_report(y_test, y_pred_tuned, output_dict=True)['1']['recall']:.2f}")
print(f"  - This metric indicates that the model correctly identifies {classification_report(y_test, y_pred_tuned, output_dict=True)['1']['recall']:.2f}% of all actual responders.")
print(f"- F1-Score for the minority class (responders, class 1): {classification_report(y_test, y_pred_tuned, output_dict=True)['1']['f1-score']:.2f}")
print(f"  - This is the harmonic mean of precision and recall for the minority class, providing a single metric that balances both.")

print(f"\n- ROC AUC Score: {roc_auc:.2f}")
print("  - The AUC represents the model's ability to distinguish between the positive class (responders) and the negative class (non-responders).")
print(f"  - An AUC of {roc_auc:.2f} suggests that the model has a good ability to discriminate between the two classes, better than random chance (AUC = 0.5).")

print("\nBusiness Implications:")
print("- High Recall is important for identifying as many potential responders as possible to maximize the campaign's reach among the target audience.")
print("- High Precision is important for minimizing marketing costs by not targeting too many non-responders.")
print("- There is often a trade-off between precision and recall. The F1-score provides a balance between these two metrics.")
print("- Accuracy alone would be misleading in this imbalanced scenario, as a model predicting no one responds would have high accuracy (95%) but would be useless for the campaign.")
print("- The ROC AUC provides a robust measure of the model's overall discriminative power, independent of the classification threshold.")

Interpretation of Evaluation Metrics:

Classification Report:
              precision    recall  f1-score   support

           0       0.98      0.86      0.92      1868
           1       0.27      0.73      0.40       132

    accuracy                           0.85      2000
   macro avg       0.63      0.80      0.66      2000
weighted avg       0.93      0.85      0.88      2000


ROC AUC Score: 0.8467206865226138

Analysis in the context of the e-commerce marketing campaign:
- Precision for the minority class (responders, class 1): 0.27
  - This metric indicates that when the model predicts a customer will respond, it is correct 0.27% of the time.
- Recall for the minority class (responders, class 1): 0.73
  - This metric indicates that the model correctly identifies 0.73% of all actual responders.
- F1-Score for the minority class (responders, class 1): 0.40
  - This is the harmonic mean of precision and recall for the minority class, providing a single metric that balances bot